zerver/webhooks/slack/view.py

```diff
@@ -1,62 +1,237 @@
+    }
+
+    # Handle Slacks "challenge" handshake when first registering endpoint
+    # to Event API.
+    if outer_data["type"] == "url_verification" and outer_data["challenge"]:
         check_send_webhook_message(
             request,
             user_profile,
-            topic_name,
-            content,
-            stream=channel_name,
+            "Integration notification",
+            "Successfully registered endpoint to Slack!",
         )
-    else:
-        raise JsonableError(_("Error: channels_map_to_topics parameter other than 0 or 1"))
+        return json_slack_challenge_response(outer_data["challenge"])
+
+    # Prevent any Zulip messages sent through the Slack Bridge from looping
+    # back here.
+    if is_bot_message(payload):
+        return json_success(request)
+
+    check_token_access(slack_app_token)
+
+    event_dict = payload.get("event", {})
+    file_dict = event_dict.get("files")
+    raw_text = event_dict.get("text").tame(check_string)
+
+    event_data: Dict[str, Any] = {
+        "type": event_dict.get("type").tame(check_string),
+        "text": convert_to_zulip_markdown(raw_text, slack_app_token),
+        "channel": get_channel_name_str(payload, slack_app_token),
+        "message_sender": get_sender_name_str(payload, slack_app_token),
+        "file_obj_list": convert_file_dict(file_dict) if file_dict else [],
+    }
+
+    if event_data["type"] == "message":
+        content = get_message_body(event_data)
+        topic_name = "Message from Slack"
+
+        if channels_map_to_topics is None:
+            check_send_webhook_message(request, user_profile, topic_name, content)
+        elif channels_map_to_topics == VALID_OPTIONS["SHOULD_BE_MAPPED"]:
+            # If the webhook URL has a user_specified_topic,
+            # then this topic-channel mapping will not be used.
+            topic_name = f"channel: {event_data['channel']}"
+            check_send_webhook_message(
+                request,
+                user_profile,
+                topic_name,
+                content,
+            )
+        elif channels_map_to_topics == VALID_OPTIONS["SHOULD_NOT_BE_MAPPED"]:
+            # This channel-channel mapping will be used even if
+            # there is a channel specified in the webhook URL.
+            check_send_webhook_message(
+                request,
+                user_profile,
+                topic_name,
+                content,
+                stream=event_data["channel"],
+            )
+        else:
+            # We want to becareful and always respond with HTTP 200 because Slack
+            # will retry sending the payload again and again if we fail to respond
+            # with 200 ok.
+
+            pass
+
+            # TODO: As a follow up might want to send a notification message
+            # informing the user about this kind of error in a user-friendly way
+            # instead of logging the error or raising Exception.
```

Comment:
FOLLOW-UP: Better handling of these kinds of errors could potentially improve UX when trying to debug why their integration doesn't work.

Reply:
Add back jsonable error